### PR TITLE
Route/Handler TypeScript normalization

### LINKED ISSUE #2511: RouteHandlerCallbackOptions type is not conform to the doc
**Library Affected**:
workbox-core (types exported)

**Browser & Platform**:
N/A

**Issue or Feature Request Description**:
According to the doc (https://developers.google.com/web/tools/workbox/reference-docs/latest/module-workbox-routing#~handlerCallback), options should be:

```js
export interface RouteHandlerCallbackOptions {
  request: Request;
  url?: URL;
  event?: FetchEvent;
  params?: string[] | {[paramName: string]: string};
}
```

but the type exported by workbox-core are https://github.com/GoogleChrome/workbox/blob/04ba6442c466d2e8197fe586672143d201af3a61/packages/workbox-core/src/types.ts#L36-L44

This type is too loose (`ExtendableEvent` instead of `FetchEvent` and `Request | string` instead of `Request`).

This makes it hard to use for example :

```js
const saveHandler: RouteHandler = ({ url, event, params }) => {
    let initid = params[1];
    let viewId = params[2];
    const workingRequest = event.request.clone();
```

which triggers `TS2339: Property 'request' does not exist on type 'ExtendableEvent'.`

neither

```js
  const saveHandler: RouteHandler = ({ url, event, params , request}) => {
    let initid = params[1];
    let viewId = params[2];
    const workingRequest = request.clone();
```

which triggers `TS2339: Property 'clone' does not exist on type 'string | Request'.   Property 'clone' does not exist on type 'string'.`

### PR DESCRIPTION
**Pull request by jeffposnick**
R: @philipwalton

Fixes #2511

I am not 100% confident about the changes in this PR, especially as it relates to the new `workbox-strategy` code in v6. So your close attention and feedback is appreciated, @philipwalton. It will be good to shake this out during the v6 pre-releases to make sure nothing unexpected was done.

There are a number of JSDoc changes that need to be made along with these TypeScript changes, and I'll do them in a follow-up commit once we're happy with all the TypeScript types.

As part of this PR, I dropped the "shadow" types that were re-exported in `workbox-routing/src/_types.ts` and just refer to everything from a single source of truth, `workbox-core/src/types.ts`.

I tested this manually with some sample usage scenarios to confirm that `tsc` didn't flag any problems. All of the following worked fine:

```ts
registerRoute(
  /\.css$/,
  ({url}) => fetch(url.href),
);

const myHandler = ({request} : {request: Request}) => {
  return fetch(request);
};
registerRoute(
  ({url}) => url.pathname.endsWith('.js'),
  myHandler,
);

const networkFirst = new NetworkFirst();
registerRoute(
  '/abcd',
  networkFirst,
);

async function testHandler(request: Request) {
  const response = await networkFirst.handle({request});
  return response;
}
registerRoute(
  '/abcd',
  ({request}) => testHandler(request),
);

async function anotherTestHandler() {
  const response = await networkFirst.handle({request: '/blah'});
  return response;
}
registerRoute(
  '/abcd',
  anotherTestHandler,
);

const saveHandler = async ({url, event, params} : {
  url: URL, event: FetchEvent, params?: string[] | {},
}) => {
  console.log(url, event, params);
  return new Response('blah');
};
registerRoute(
  '/abcd',
  saveHandler,
);
```

### CODE REVIEW

File: packages/workbox-core/src/types.ts [hunk 1]
```diff
@@ -48,12 +48,23 @@ export interface RouteMatchCallback {
...
+  url: URL;
   params?: string[] | MapLikeObject;
+}
+
+/**
+ * Options passed to a `ManualHandlerCallback` function.
+ */
+export interface ManualHandlerCallbackOptions {
+  request: Request | string;
   event?: ExtendableEvent;
```
Comment (philipwalton) [inline packages/workbox-core/src/types.ts]: I believe we want to make this required in v6.

File: packages/workbox-core/src/types.ts [hunk 2]
```diff
@@ -48,12 +48,23 @@ export interface RouteMatchCallback {
...
+  event: ExtendableEvent;
+  request: Request;
+  url: URL;
   params?: string[] | MapLikeObject;
+}
+
+/**
+ * Options passed to a `ManualHandlerCallback` function.
+ */
+export interface ManualHandlerCallbackOptions {
```
Comment (philipwalton) [inline packages/workbox-core/src/types.ts:60]: Could `RouteHandlerCallbackOptions` extend `ManualHandlerCallbackOptions` (especially if we make `event` required)?

File: packages/workbox-core/src/types.ts [hunk 3]
```diff
@@ -66,6 +77,18 @@ export interface RouteHandlerCallback {
...
 
+/**
+ * The "handler" callback is invoked whenever a `Router` matches a URL/Request
+ * to a `Route` via its `RouteMatchCallback`. This handler callback should
+ * return a `Promise` that resolves with a `Response`.
+ *
+ * If a non-empty array or object is returned by the `RouteMatchCallback` it
+ * will be passed in as this handler's `options.params` argument.
+ */
+export interface HandlerCallback {
```
Comment (philipwalton) [inline packages/workbox-core/src/types.ts]: This looks identical to the `RouteHandlerCallback` type above, and it doesn't seem to be used anywhere in the code base. Is it needed?

File: packages/workbox-routing/src/_types.ts [hunk 4]
```diff
@@ -6,25 +6,8 @@
...
-  RouteHandlerObject as HandlerObject,
-  RouteHandlerCallback as HandlerCallback,
-  RouteHandlerCallbackOptions as HandlerCallbackOptions,
-  RouteMatchCallback as MatchCallback,
-  RouteMatchCallbackOptions as MatchCallbackOptions,
-}
-
 // * * * IMPORTANT! * * *
 // ------------------------------------------------------------------------- //
 // jdsoc type definitions cannot be declared above TypeScript definitions or
```
Comment (philipwalton) [inline packages/workbox-routing/src/_types.ts:13]: This was a problem before, but it looks like the jsdoc comments in this file are no longer up-to-date. Can you fix that as part of this PR?

File: packages/workbox-strategies/src/Strategy.ts [hunk 5]
```diff
@@ -115,7 +110,7 @@ abstract class Strategy implements RouteHandlerObject {
    * @param {URL} [options.url]
    * @param {*} [options.params]
    */
-  handle(options: FetchEvent | RouteHandlerCallbackOptions): Promise<Response> {
+  handle(options: HandlerCallbackOptions): Promise<Response> {
```
Comment (philipwalton) [inline packages/workbox-strategies/src/Strategy.ts]: I think I'd still like to have this to accept a `FetchEvent` object, since that's how it's most often going to be used:

```js
addEventListener('fetch', (event) => {
   event.respondWith(someStrategy.handle(event));
});
```

I remember in the past you expressed a preference for `handle({event})` (using object shorthand), but that doesn't work if we're going to require both the `request` and the `event` param (which I think we want to?).

WDYT?

File: packages/workbox-strategies/src/Strategy.ts [hunk 6]
```diff
@@ -150,12 +145,14 @@ abstract class Strategy implements RouteHandlerObject {
...
-      // `options.request` can be a string, similar to what `fetch()` accepts.
-      options.request = new Request(options.request);
     }
 
-    const {event, request, params} = options as StrategyHandlerOptions;
+    const event = options.event;
+    const request = typeof options.request === 'string' ?
+        new Request(options.request) :
+        options.request;
+    const params = 'params' in options ? options.params : undefined;
```
Comment (philipwalton) [inline packages/workbox-strategies/src/Strategy.ts:158]: Why the change away from destructuring here?

File: packages/workbox-core/src/types.ts [hunk 7]
```diff
@@ -48,12 +48,23 @@ export interface RouteMatchCallback {
...
+
+/**
+ * Options passed to a `ManualHandlerCallback` function.
+ */
+export interface ManualHandlerCallbackOptions {
+  request: Request | string;
   event?: ExtendableEvent;
 }
 
+export type HandlerCallbackOptions =
```
Comment (philipwalton) [inline packages/workbox-core/src/types.ts:65]: Similar question to above. Would it work to make something like `HandlerCallbackOptions` a base type, and then any modules that need to accept `RouteHandlerCallbackOptions | ManualHandlerCallbackOptions` could accept any subtype of `HandlerCallbackOptions`?

**Review by philipwalton — CHANGES_REQUESTED** [review]

File: packages/workbox-core/src/types.ts (hunk 1 above)
Comment (jeffposnick) [reply packages/workbox-core/src/types.ts]: Oh, right. We had talked about how using strategies manually outside of an event handler wasn't a great idea.

File: packages/workbox-core/src/types.ts (hunk 2 above)
Comment (jeffposnick) [reply packages/workbox-core/src/types.ts:60]: Yup.

File: packages/workbox-core/src/types.ts (hunk 1 above)
Comment (jeffposnick) [reply packages/workbox-core/src/types.ts]: One follow-up from that is that I need to change a lot of the parameter signatures in the `PrecacheController` class to mark the `ExtendableEvent` as being required. (That's probably a good thing, since using `PrecacheController` outside of an event sounds like a recipe for disaster.)

Comment (jeffposnick) [reply packages/workbox-core/src/types.ts]: This change also filters through to the `MatchWrapperOptions` and `PutWrapperOptions` interfaces, which now need to mark the `event` as required.

File: packages/workbox-core/src/types.ts (hunk 2 above)
Comment (jeffposnick) [reply packages/workbox-core/src/types.ts:60]: Actually, no, because `ManualHandlerCallbackOptions` defines `request: Request | string;` and we want `RouteHandlerCallbackOptions` to be `request: Request;`.

There are enough different about the two that it would be easier to keep them separate rather than extend and override.

File: packages/workbox-core/src/types.ts (hunk 3 above)
Comment (jeffposnick) [reply packages/workbox-core/src/types.ts]: So my intention here was to export an interface that developers could implement when they wrote their own "manual" handler functions. But `HandlerCallback `, as implemented, is probably too generic, since it could take either `RouteHandlerCallbackOptions` or `ManualHandlerCallbackOptions` as its parameter.

I've gotten more specific by removing `HandlerCallback` and exporting an explicit `ManualHandlerCallback` instead, that takes `ManualHandlerCallbackOptions`.

So even though `ManualHandlerCallbackOptions` isn't used by the codebase, developers could take advantage of it via

```
import {ManualHandlerCallback} from 'workbox-core/types';

const myHandler: ManualHandlerCallback = async({request, event}) => {
  const networkFirst = new NetworkFirst();
  const networkResponse = networkFirst.handle({event, request});
  return networkResponse || new Response('...');
};
```

File: packages/workbox-routing/src/_types.ts (hunk 4 above)
Comment (jeffposnick) [reply packages/workbox-routing/src/_types.ts:13]: Once the actual types are 👍 , I'll go back and adjust the JSDocs. It's going to be too much wasted effort otherwise.

File: packages/workbox-strategies/src/Strategy.ts (hunk 5 above)
Comment (jeffposnick) [reply packages/workbox-strategies/src/Strategy.ts]: Yeah, I'm adding that `FetchEvent` back as one of the supported types, especially since the code supports it.

File: packages/workbox-strategies/src/Strategy.ts (hunk 6 above)
Comment (jeffposnick) [reply packages/workbox-strategies/src/Strategy.ts:158]: `HandlerCallbackOptions` may not contain `params` (`RouteHandlerCallbackOptions` will, but `ManualHandlerCallbackOptions` won't), and explicitly checking instead of just using destructing was necessary to get TypeScript happy.

File: packages/workbox-core/src/types.ts (hunk 7 above)
Comment (jeffposnick) [reply packages/workbox-core/src/types.ts:65]: See my earlier comment about wanting to keep `RouteHandlerCallbackOptions` and `ManualHandlerCallbackOptions` independent. Making `HandlerCallbackOptions` a composite of both of them seems to work for our use case.

**Comment by jeffposnick** [comment]
PTAL.

File: packages/workbox-core/src/types.ts (hunk 2 above)
Comment (philipwalton) [reply packages/workbox-core/src/types.ts:60]: Gotcha.

File: packages/workbox-core/src/types.ts (hunk 3 above)
Comment (philipwalton) [reply packages/workbox-core/src/types.ts]: SGTM

File: packages/workbox-routing/src/_types.ts (hunk 4 above)
Comment (philipwalton) [reply packages/workbox-routing/src/_types.ts:13]: SGTM

File: packages/workbox-strategies/src/Strategy.ts (hunk 5 above)
Comment (philipwalton) [reply packages/workbox-strategies/src/Strategy.ts]: Cool.

File: packages/workbox-core/src/types.ts (hunk 7 above)
Comment (philipwalton) [reply packages/workbox-core/src/types.ts:65]: SGTM.

File: packages/workbox-strategies/src/Strategy.ts (hunk 6 above)
Comment (philipwalton) [reply packages/workbox-strategies/src/Strategy.ts:158]: Kk

File: packages/workbox-core/src/types.ts (hunk 1 above)
Comment (philipwalton) [reply packages/workbox-core/src/types.ts]: Not sure if you want to do this as part of this PR or a separate on, but there are many cases in our code where we've had to make an event optional in a type definition due to us not making it required in handlers (e.g. [here](https://github.com/GoogleChrome/workbox/blob/f6cecf266436b4b199d6a80ece4f09a8515609d0/packages/workbox-strategies/src/StrategyHandler.ts#L41)). I did a in VSCode for `event?:`, and lots of cases come up (and there are probably others where the name "event" isn't used).

Also, there are cases where we're conditionally checking whether an event was passed (e.g. [here](https://github.com/GoogleChrome/workbox/blob/f6cecf266436b4b199d6a80ece4f09a8515609d0/packages/workbox-strategies/src/StrategyHandler.ts#L120-L122)), and it probably makes sense to remove those now.

How do you want to handle these?

Comment (jeffposnick) [reply packages/workbox-core/src/types.ts]: Okay, I can get those updated in this PR as well.

**Review by philipwalton — APPROVED** [review]

File: packages/workbox-core/src/types.ts (hunk 1 above)
Comment (philipwalton) [reply packages/workbox-core/src/types.ts]: Thanks, I went through and updated the corresponding jsdoc comments for a few that where missed (it was easier than going back and forth).

**Comment by jeffposnick** [comment]
Okay! I think with your JSDoc updates, @philipwalton, and all the tests now passing, that's it for this PR.

If there's anything I left out, we can pick it up prior to the v6 release, but I'm merging for now.
